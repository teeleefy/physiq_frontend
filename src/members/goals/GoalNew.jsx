import { Form, FormGroup, Input, Button, Label, Alert } from "reactstrap";
import { useState, useContext, useEffect } from "react";
import {MemberContext} from "../../auth/UserContext.js";
import PhysiqApi from "../../Api.js";
import { useParams, useNavigate } from "react-router-dom";
import Loading from "../../navigation/Loading.jsx";
import '../styles/Goal.css'

function GoalNew(){
    let navigate = useNavigate();
    let { goalId } = useParams();
    const { currentMember } = useContext(MemberContext);
    const [currentGoal, setCurrentGoal] = useState([]);
    const [isLoading, setIsLoading] = useState(true);
    const [formMessages, setFormMessages] = useState([]);
    const [updateSuccess, setUpdateSuccess] =useState();
    const [formData, setFormData] = useState({goalName: "", goalDetails: ""});


    async function addGoal() {
        try {
          let newGoal= await PhysiqApi.addMemberGoal(currentMember.id, formData);
          return { success: true };
        } catch (errors) {
          console.error("Add Goal failed", errors);
          return { success: false, errors };
        }
      }

    async function handleSubmit(evt){
        evt.preventDefault();
        // console.log(formData);
        let result = await addGoal();
        
        if(result.success){
        //    alert("Saved Changes!") 
           setFormMessages(['Goal Added!'])
           setUpdateSuccess(true);
           navigate("..", { relative: "path"});
        }
        else{
            setFormMessages(result.errors);
            setUpdateSuccess(false);
        }
    };

  
    /** Update local state w/curr state of input elem */
  
    const handleChange = evt => {
      const { name, value }= evt.target;
      setFormData(fData => ({
        ...fData,
        [name]: value
      }));
    };

    return(
        <>
            <Form className="GoalNew-Form m-4" >
            <h1 className="Goal-h1">Add Goal:</h1>
            <p className="text-secondary">Fields marked with <span className="text-danger">*</span> are required.</p>
            <FormGroup>
                <Label className="Goal-label" for="goalName" >
                    <b>Goal Name </b><span className="text-danger">*</span>
                </Label>
                <Input
                id="goalName"
                name="goalName"
                value={formData.goalName}
                placeholder="Enter goal name..."
                type="text"
                onChange={handleChange}
                />
            </FormGroup>
            <FormGroup>
                <Label className="Goal-label" for="goalDetails">
                    <b>Details</b>
                </Label>
                <Input
                id="goalDetails"
                name="goalDetails"
                value={formData.goalDetails}
                placeholder="Enter goal details..."
                type="text"
                onChange={handleChange}
                />
                </FormGroup>
                {formMessages.length
                    ? formMessages.map(msg => <Alert color={updateSuccess ? "success": "danger"}>{msg}</Alert>)
                    : null
                }
                
                <Button className="btn-dark" onClick={handleSubmit}>Add Goal</Button>
            </Form>
        </>
    )}
    
export default GoalNew;