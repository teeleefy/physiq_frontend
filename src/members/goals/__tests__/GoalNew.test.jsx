import { renderWithRouterAndMemberContext, screen } from '../../../../__tests__/test-tools';
import GoalNew from '../GoalNew'


describe('Goal New', () => {
  it('renders the GoalNew component', () => {
    renderWithRouterAndMemberContext(<GoalNew />)
    
    screen.debug(); // prints out the jsx in the GoalNew component unto the command line
  })

  it("matches snapshot", function() {
    const { asFragment } = renderWithRouterAndMemberContext(<GoalNew />);
    expect(asFragment()).toMatchSnapshot();
  });

})
