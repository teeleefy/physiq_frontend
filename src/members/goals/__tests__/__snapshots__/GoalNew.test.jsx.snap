// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Goal > matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="GoalNew-Form m-4"
  >
    <h1
      class="Goal-h1"
    >
      Add Goal:
    </h1>
    <p
      class="text-secondary"
    >
      Fields marked with 
      <span
        class="text-danger"
      >
        *
      </span>
       are required.
    </p>
    <div
      class="mb-3"
    >
      <label
        class="Goal-label form-label"
        for="goalName"
      >
        <b>
          Goal Name 
        </b>
        <span
          class="text-danger"
        >
          *
        </span>
      </label>
      <input
        class="form-control"
        id="goalName"
        maxlength="100"
        name="goalName"
        placeholder="Enter goal name..."
        type="text"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="Goal-label form-label"
        for="goalDetails"
      >
        <b>
          Details
        </b>
      </label>
      <textarea
        class="form-control"
        id="goalDetails"
        maxlength="250"
        name="goalDetails"
        placeholder="Enter goal details..."
      />
    </div>
    <p
      class="text-secondary"
    >
      0/250 Characters
    </p>
    <button
      class="btn-dark btn btn-secondary"
      type="button"
    >
      Add Goal
    </button>
  </form>
</DocumentFragment>
`;

exports[`Goal New > matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="GoalNew-Form m-4"
  >
    <h1
      class="Goal-h1"
    >
      Add Goal:
    </h1>
    <p
      class="text-secondary"
    >
      Fields marked with 
      <span
        class="text-danger"
      >
        *
      </span>
       are required.
    </p>
    <div
      class="mb-3"
    >
      <label
        class="Goal-label form-label"
        for="goalName"
      >
        <b>
          Goal Name 
        </b>
        <span
          class="text-danger"
        >
          *
        </span>
      </label>
      <input
        class="form-control"
        id="goalName"
        maxlength="100"
        name="goalName"
        placeholder="Enter goal name..."
        type="text"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="Goal-label form-label"
        for="goalDetails"
      >
        <b>
          Details
        </b>
      </label>
      <textarea
        class="form-control"
        id="goalDetails"
        maxlength="250"
        name="goalDetails"
        placeholder="Enter goal details..."
      />
    </div>
    <p
      class="text-secondary"
    >
      0/250 Characters
    </p>
    <button
      class="btn-dark btn btn-secondary"
      type="button"
    >
      Add Goal
    </button>
  </form>
</DocumentFragment>
`;
